src/lss.jl

```diff
@@ -0,0 +1,422 @@
+  end
+  return nothing
+end
+
+function wB!(S::LSSSchur,Δt,Nt,numindvar,dt)
+  @unpack wBinv = S
+  fill!(wBinv, one(Δt))
+  dim = numindvar * Nt
+  tmp = @view wBinv[1:numindvar]
+  tmp ./= dt[1]
+  tmp = @view wBinv[dim-2:end]
+  tmp ./= dt[end]
+  for indx = 2:Nt-1
+    tmp = @view wBinv[(indx-1)*numindvar+1:indx*numindvar]
+    tmp ./= (dt[indx]+dt[indx-1])
+  end
+
+  wBinv .*= 2*Δt
+  return nothing
+end
+
+wE!(S::LSSSchur,Δt,dt,alpha::Union{CosWindowing,Cos2Windowing}) = nothing
+
+function wE!(S::LSSSchur,Δt,dt,alpha)
+  @unpack wEinv = S
+  @. wEinv = Δt/(alpha^2*dt)
+  return nothing
+end
+
+function B!(S::LSSSchur,dt,umid,sense,sensealg)
+  @unpack B = S
+  @unpack f,J,uf,numindvar,f_cache,jac_config = sense
+
+  fill!(B, zero(eltype(J)))
+
+  for (i,u) in enumerate(eachcol(umid))
+    if DiffEqBase.has_jac(f)
+      f.jac(J,u,uf.p,uf.t) # Calculate the Jacobian into J
+    else
+      jacobian!(J, uf, u, f_cache, sensealg, jac_config)
+    end
+    B0 = @view B[(i-1)*numindvar+1:i*numindvar,i*numindvar+1:(i+1)*numindvar]
+    B1 =  @view B[(i-1)*numindvar+1:i*numindvar,(i-1)*numindvar+1:i*numindvar]
+    B0 .+= I/dt[i] - J/2
+    B1 .+= -I/dt[i] -J/2
+  end
+  return nothing
+end
+
+E!(S::LSSSchur,dudt,alpha::Union{CosWindowing,Cos2Windowing}) = nothing
+
+function E!(S::LSSSchur,dudt,alpha)
+  @unpack E = S
+  numindvar, Ndt = size(dudt)
+  for i=1:Ndt
+    tmp = @view E[(i-1)*numindvar+1:i*numindvar,i]
+    copyto!(tmp, (@view dudt[:,i]))
+  end
+  return nothing
+end
+
+# compute Schur
+function S!(S::LSSSchur)
+  @unpack B, E, wBinv, wEinv, Smat = S
+  Smat .= B*Diagonal(wBinv)*B'
+  (wEinv !== nothing) && (Smat .+= E*Diagonal(wEinv)*E')
+  return nothing
+end
+
+function b!(b, prob::ForwardLSSProblem)
+  @unpack diffcache, umid, sensealg = prob
+  @unpack f, f_cache, pJ, pf, paramjac_config, uf, numindvar = diffcache
+
+  for (i,u) in enumerate(eachcol(umid))
+    if DiffEqBase.has_paramjac(f)
+      f.paramjac(pJ, u, uf.p, pf.t)
+    else
+      pf.u = u
+      jacobian!(pJ, pf, uf.p, f_cache, sensealg, paramjac_config)
+    end
+    tmp = @view b[(i-1)*numindvar+1:i*numindvar,:]
+    copyto!(tmp, pJ)
+  end
+  return nothing
+end
+
+function solve(prob::ForwardLSSProblem; t0skip=zero(prob.Δt), t1skip=zero(prob.Δt))
+  _solve(prob,prob.sensealg,prob.sensealg.alpha,t0skip,t1skip)
+end
+
+function _solve(prob::ForwardLSSProblem,sensealg::ForwardLSS,alpha::Number,t0skip,t1skip)
+  @unpack sol, S, window, Δt, diffcache, b, w, v, η, res, g, g0, umid = prob
+  @unpack wBinv, wEinv, B, E, Smat = S
+  @unpack dg_val, pgpu, pgpu_config, pgpp, pgpp_config, numparams, numindvar, uf = diffcache
+
+
+  n0 = searchsortedfirst(sol.t, sol.t[1]+t0skip)
+  n1 = searchsortedfirst(sol.t, sol.t[end]-t1skip)
+
+  b!(b,prob)
+
+  ures = @view sol.u[n0:n1]
+  umidres = @view umid[:,n0:n1-1]
+
+  # reset
+  res .*=false
+
+  for i=1:numparams
+    #running average
+    g0 *= false
+    bpar = @view b[:,i]
+    w .= Smat\bpar
```

Comment:
Smat is constant in the loop right? Then you will want to hoist out `_Smat = lu(Smat)` so it's only factorized once.

Reply:
yep, Smat is constant. Sounds good!